### PR TITLE
[Enhancement] MessageHub.get_info() supports returning a default value

### PR DESCRIPTION
**Pull request by enkilee**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Reslove #988

## Modification

Reslove #988

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/logging/message_hub.py [hunk 1]
```diff
@@ -296,22 +296,23 @@ def get_scalar(self, key: str) -> HistoryBuffer:
                            f'instance name is: {MessageHub.instance_name}')
         return self.log_scalars[key]
 
-    def get_info(self, key: str) -> Any:
-        """Get runtime information by key.
+    def get_info(self, key: str, default: OrderedDict) -> Any:
```
Comment (HAOCHENYE) [inline mmengine/logging/message_hub.py:299]: Hi, this default value should be optional. The intention behind adding the default value is to enable `get_info` to behave like the `dict.get()`, which can return a user-specified default value when the key does not exist.

File: mmengine/logging/message_hub.py [hunk 2]
```diff
@@ -296,22 +296,23 @@ def get_scalar(self, key: str) -> HistoryBuffer:
...
+        """Get runtime information by key. if.
```
Comment (HAOCHENYE) [inline mmengine/logging/message_hub.py:300]: Please complete the docstring

File: mmengine/logging/message_hub.py [hunk 3]
```diff
@@ -296,22 +296,25 @@ def get_scalar(self, key: str) -> HistoryBuffer:
                            f'instance name is: {MessageHub.instance_name}')
         return self.log_scalars[key]
 
-    def get_info(self, key: str) -> Any:
-        """Get runtime information by key.
+    def get_info(self, key: str, default: Optional[OrderedDict] = None) -> Any:
```
Comment (HAOCHENYE) [inline mmengine/logging/message_hub.py:299]:
```suggestion
    def get_info(self, key: str, default: Optional[Any] = None) -> Any:
```

File: mmengine/logging/message_hub.py [hunk 4]
```diff
@@ -296,22 +296,25 @@ def get_scalar(self, key: str) -> HistoryBuffer:
...
+        """Get runtime information by key. if key does not exists, this method
+        will return default information.
```
Comment (HAOCHENYE) [inline mmengine/logging/message_hub.py:301]:
```suggestion
        """Get runtime information by key. if the key does not exist, this method
        will return default information.
```

File: mmengine/logging/message_hub.py [hunk 5]
```diff
@@ -296,22 +296,25 @@ def get_scalar(self, key: str) -> HistoryBuffer:
...
 
         Args:
             key (str): Key of runtime information.
+            default (OrderedDict, optional): Each key in the dictionary is the
+            name of the information. Defaults to None.
```
Comment (HAOCHENYE) [inline mmengine/logging/message_hub.py:306]:
```suggestion
            default (OrderedDict, optional): The default returned value for the given key.
```

File: mmengine/logging/message_hub.py [hunk 6]
```diff
@@ -296,22 +296,25 @@ def get_scalar(self, key: str) -> HistoryBuffer:
...
         """
         if key not in self.runtime_info:
-            raise KeyError(f'{key} is not found in Messagehub.log_buffers: '
-                           f'instance name is: {MessageHub.instance_name}')
-
-        # TODO： There are restrictions on objects that can be saved
-        # return copy.deepcopy(self._runtime_info[key])
-        return self._runtime_info[key]
+            default = self._runtime_info
+            return default
```
Comment (HAOCHENYE) [inline mmengine/logging/message_hub.py:313]:
```suggestion
            if default is not None:
                return default
            else:
               raise KeyError(f'{Can not find {key} in runtime information of message_hub')
```

**Comment by HAOCHENYE** [comment]
Hi! The CI seems failed!

File: tests/test_logging/test_message_hub.py [hunk 7]
```diff
@@ -83,7 +83,7 @@ def test_get_scalar(self):
     def test_get_runtime(self):
         message_hub = MessageHub.get_instance('mmengine')
         with pytest.raises(KeyError):
-            message_hub.get_info('unknown')
+            message_hub.get_info('unknown', ['test_value'])
```
Comment (HAOCHENYE) [inline tests/test_logging/test_message_hub.py:86]: Why will this line raise a `KeyError`?

**Comment by HAOCHENYE** [comment]
Hi, the CI seems failed, please fix it

image

**Comment by enkilee** [comment]
> Hi, the CI seems failed, please fix it
> 
> image
> 

ok
pc is using for remoter, i will fix it asap

File: mmengine/logging/message_hub.py [hunk 8]
```diff
@@ -296,22 +296,29 @@ def get_scalar(self, key: str) -> HistoryBuffer:
...
-
-        # TODO： There are restrictions on objects that can be saved
-        # return copy.deepcopy(self._runtime_info[key])
-        return self._runtime_info[key]
+            if default is not None:
+                return default
+            else:
+                raise KeyError(
+                    f'{key} is not found in Messagehub.log_buffers: '
+                    f'instance name is: {MessageHub.instance_name}')
```
Comment (HAOCHENYE) [inline mmengine/logging/message_hub.py:317]:
```suggestion
            return default
```

File: tests/test_logging/test_message_hub.py [hunk 9]
```diff
@@ -186,11 +184,6 @@ def test_getstate(self):
         obj = pickle.dumps(message_hub)
         instance = pickle.loads(obj)
 
-        with pytest.raises(KeyError):
-            instance.get_info('feat')
-        with pytest.raises(KeyError):
-            instance.get_info('lr')
```
Comment (HAOCHENYE) [inline tests/test_logging/test_message_hub.py:192]: We should also test the returned value is None.

**Comment by enkilee** [comment]
@HAOCHENYE  hi, I've a question about "pytest.raises()" ,if  test the returned value is None, pytest.raises(?)

File: tests/test_logging/test_message_hub.py [hunk 10]
```diff
@@ -82,8 +82,8 @@ def test_get_scalar(self):
 
     def test_get_runtime(self):
         message_hub = MessageHub.get_instance('mmengine')
-        with pytest.raises(KeyError):
-            message_hub.get_info('unknown')
+        tgrresult = message_hub.get_info('unknown')
+        assert tgrresult is None
```
Comment (zhouzaida) [inline tests/test_logging/test_message_hub.py:86]:
```suggestion
        assert message_hub.get_info('unknown') is None
```

File: tests/test_logging/test_message_hub.py [hunk 11]
```diff
@@ -186,10 +186,10 @@ def test_getstate(self):
...
         instance = pickle.loads(obj)
 
-        with pytest.raises(KeyError):
-            instance.get_info('feat')
-        with pytest.raises(KeyError):
-            instance.get_info('lr')
+        featresult = instance.get_info('feat')
+        assert featresult is None
+        lrresult = instance.get_info('lr')
+        assert lrresult is None
```
Comment (zhouzaida) [inline tests/test_logging/test_message_hub.py:192]:
```suggestion
        assert instance.get_info('feat') is None
        assert instance.get_info('lr') is None
```

**Review by zhouzaida — DISMISSED** [review]

**Review by zhouzaida — APPROVED** [review]
